Downgrade minimal required python to 3.11

clane9 opened the pull request:
Requiring python>=3.12 is a bit of a burden. It's only required for `find_bids_datasets`, which uses `Path.walk`. Instead, downgrade minimum python to 3.11 and add a guard on this function in case python<3.12.

Note, we could look into downgrading further. The next block would be `get_column_names` returns a `StrEnum`, which was introduced in 3.11. I'm hesitant to remove this though, because being able to treat these enum fields as native strings is nice.

clane9 committed:
Downgrade minimal required python to 3.11

Requiring python>=3.12 is a bit of a burden. It's only required for
`find_bids_datasets`, which uses `Path.walk`. Instead, downgrade minimum
python to 3.11 and add a guard on this function in case python<3.12.

Note, we could look into downgrading further. The next block would be
`get_column_names` returns a `StrEnum`, which was introduced in 3.11.
I'm hesitant to remove this though, because being able to treat these
enum fields as native strings is nice.

effigies commented:
I would recommend not dropping old versions of dependencies any faster than https://scientific-python.org/specs/spec-0000/.

clane9 commented:
Thanks for the link @effigies, that's a useful reference. So by that rule, it seems ok to not support 3.10, but indeed we should support 3.11.

kaitj reviewed (commented):
Code-wise this largely looks good to me. 

I did play around with this briefly to try to get s3 datasets working with py311, though had to change `_indexing.py` code code (namely using `os.walk` instead of `path.walk`) + type casting a little bit to get the tests passing. That said, I'm not sure what gets returned or if it actually works with `b2t2 find` - didn't dig too deep into this yet. Could leave this as a todo down the road + add in matrix for testing for supported py versions.

_I've added the s3 support in py311 in #52._

kaitj committed:
Switch path.walk to os.walk for py311

- Enables use of `finds_bids_datasets` in py311.
  - `root` in `_indexing.py` is initially passed as original type, with walked `dirpath` typecasted as Path
- Removed error for <py312 in throughout codebase + testing

kaitj committed:
Setup python matrix for testing

- Only runs after formatting
- Update to dependencies to include other versions of python support

clane9 committed:
Implement iterative directory walk for find

`Path.walk` and `CloudPath.walk` depend on python>=3.12. Also,
`CloudPath.walk` retrieves all files up front rather than iteratively.
Here we add some directory walk logic of our own for iteratively finding
BIDS datasets under a root directory.

clane9 committed:
Update README.md

clane9 commented:
I decided to just copy over the basic directory walk logic from `os.walk`. This removes the requirement for python 3.12, as well as any awkward guard on `find_bids_datasets` in case python<3.12. Also, this way we can yield discovered cloud datasets iteratively, rather than have to [wait for a full scan of the cloud directory](https://github.com/drivendataorg/cloudpathlib/blob/v0.21.1/cloudpathlib/cloudpath.py#L573).

@kaitj would you mind taking another look?

clane9 committed:
Update module docs

kaitj reviewed (approved):
Looks good to me - feels faster to me, but maybe I am imagining it 😅 
Had a few questions, but nothing blocking.

clane9 commented on bids2table/_indexing.py:
```diff
@@ -129,51 +129,66 @@ def get_column_names() -> enum.StrEnum:
     exclude: str | list[str] | None = None,
-    follow_symlinks: bool = True,
-    log_frequency: int = 100,
+    maxdepth: int | None = None,
```
Yep `Path.is_dir()` [follows symlinks by default](https://docs.python.org/3/library/pathlib.html#pathlib.Path.is_dir).
